python/ray/data/_internal/batcher.py

```diff
@@ -94,6 +103,10 @@ def next_batch(self) -> Block:
         if self._batch_size is None:
             assert len(self._buffer) == 1
             block = self._buffer[0]
+            if self._ensure_copy:
```

Comment:
sorry I may miss some context, why we need copy the block here?

Reply:
If we don't ensure a copy of the block/batch, and the UDF provided by the user mutates the batch (e.g. `df["new_col"] = df["old_col"] + 1`), this will fail if the underlying batch is in the Ray object store (those objects are immutable/read-only). This is preserving the existing semantics of `.map_batches()` btw, we are already doing this in master.